services/shhext/service.go

```diff
@@ -179,3 +181,46 @@ func (s *Service) Stop() error {
 
 	return s.Service.Stop()
 }
+
+func (s *Service) syncMessages(ctx context.Context, mailServerID []byte, r whisper.SyncMailRequest) (resp whisper.SyncEventResponse, err error) {
+	err = s.w.SyncMessages(mailServerID, r)
+	if err != nil {
+		return
+	}
+
+	// Wait for the response which is received asynchronously as a p2p packet.
+	// This packet handler will send an event which contains the response payload.
+	events := make(chan whisper.EnvelopeEvent, 1024)
+	sub := s.w.SubscribeEnvelopeEvents(events)
+	defer sub.Unsubscribe()
+
+	// Add explicit timeout context, otherwise the request
+	// can hang indefinitely if not specified by the sender.
+	// Sender is usually through netcat or some bash tool
+	// so it's not really possible to specify the timeout.
+	timeoutCtx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+
+	for {
+		select {
+		case event := <-events:
+			if event.Event != whisper.EventMailServerSyncFinished {
+				continue
+			}
+
+			log.Info("received EventMailServerSyncFinished event", "data", event.Data)
+
+			var ok bool
+
+			resp, ok = event.Data.(whisper.SyncEventResponse)
```

Comment:
event from concurrent sync request can be received, is it possible to compare request id?

Reply:
Not really, there are no IDs that would allow that. But also a single requester shouldn't make more than one call. I think we can assume that safely.